### ISSUE TITLE
ipywidgets.Output has stopped working in voici (still works in jupyterlite though)

### ISSUE THREAD

**Issue by LiamBindle** [voila-dashboards/voici#127]
## Description

I recently upgraded to xeus-python=0.17 because 0.16 stopped working for me and now I'm having trouble getting `ipywidgets.Output()` to work. When I update an output widget nothing happens. I think the bug is stems from voici because when I build the notebook with jupyterlite  `ipywidgets.Output()` works as expected. 

I created a reproducer [here](https://github.com/LiamBindle/voici-ipywidget-output-bug-reproducer)

```python
import ipywidgets
from datetime import datetime

output = ipywidgets.Output()
button = ipywidgets.Button(description='Click me')

display(button)
display(output)

def on_click(*args):
    output.clear_output()
    
    with output:
        print(datetime.now().isoformat())
button.on_click(on_click)
```

With jupyterlite (**expected behaviour**)
![Screenshot_1](https://github.com/user-attachments/assets/3fb94bb8-4e8b-43b0-828a-63bade6cd6db)

With voici (**nothing happens**)
![Screenshot_2](https://github.com/user-attachments/assets/e5973f17-c1f2-4b28-aa6f-1c7bcda7c82e)

## Reproduce

You can build the reproducer like so (I also included a build-environment.yml within...I'm using voici 0.7)
```
git clone https://github.com/LiamBindle/voici-ipywidget-output-bug-reproducer.git
cd voici-ipywidget-output-bug-reproducer
voici build --contents content --apps lab --apps retro
```

Serve the files
```
cd _output
python -m http.server 8010
```

See that `ipywidgets.Output()` works in jupyterlite: http://0.0.0.0:8010/lab/index.html 

See that `ipywidgets.Output()` doesn't work in voici: http://0.0.0.0:8010/voici/render/content/reproducer.html?

---

Any thoughts? Thanks in advance!

**Comment by LiamBindle** [comment]
CCing @martinRenou @DerThorsten and @jtpio

**Comment by DerThorsten** [comment]
@LiamBindle unfortunately I dont know much about voici, but hope that @martinRenou  or @jtpio can help

**Comment by LiamBindle** [comment]
Thanks @DerThorsten and @martinRenou

Hmm, strange that there should be anything unique to voici. I'm definitely seeing the bug in voici but not jupyterlite. That seems like a good lead. 

I'll try the new jupyterlite-xeus right now and get back to you

**Comment by LiamBindle** [comment]
@martinRenou I'm afraid upgrading to jupyterlite-xeus=2.0.0 didn't fix the problem. I've double checked that my build environment is indeed using jupyterlite-xeus=2.0.0. Here's a comparison of the jupyter lite vs voici console log.

<details>
<summary>Console log with Jupyter Lite (works)</summary>

![image](https://github.com/user-attachments/assets/d1a6406c-b3e3-4820-b001-bea25701baa1)

![image](https://github.com/user-attachments/assets/02b8b5bb-280b-431b-9077-e4dba511f9b7)

</details>

<details>
<summary>Console log with Voici (doesn't work)</summary>

![image](https://github.com/user-attachments/assets/df8b6e90-aa41-4491-b1ba-ddb4278084c8)

![image](https://github.com/user-attachments/assets/d09256e0-7dbc-4db0-ab16-eacb3e6c8e0e)

</details>

The main differences I see are:
1. The voici log has a bunch of warnings about unsatisfied versions <details><summary>Example</summary>

    `Unsatisfied version 4.3.4 from _JUPYTERLAB.CORE_OUTPUT of shared singleton module @jupyterlab/apputils (required ^4.3.5)` 

    </details>
2. The voici log has an undefined exception (before the xpython kernel is selected) whereas the jupyter lite log has an unexpected end of input exception.

Do you have any insight? I'm happy to keep digging but I'd benefit if someone could point me in the right direction.

**Comment by trungleduc** [comment]
@LiamBindle are you using the latest version of ipywidgets? could you try some previous versions? There are some changes in the newest version that might affect voici.

**Comment by datakurre** [comment]
@LiamBindle Have you found the reason? I probably have the same issue, found this and also tried out your reproducer with the same issue still there.

**Comment by LiamBindle** [comment]
@datakurre I'm afraid not. I spent a day or so trying to figure out what was happening, but I don't think I can dig deeper without a lead. I'm pursuing a workaround for our applications (switching to voila).

If anyone has any insight I'd be keen to dig deeper.

**Comment by datakurre** [comment]
@LiamBindle Would you have a any working builds where I could get “last known working set” of versions to try? Maybe even for the previous xeus-python?

**Comment by LiamBindle** [comment]
@datakurre I believe the previous versions have stopped working. My deployments with xeus-python=0.16 were working in July, but since then the versions that I had pinned stopped working in new deployments (errors related to failing to create default directories for Python). This prompted me to switch to xeus-python=0.17 which unveiled this issue. 

That's why it's an unfortunate situation---the current version has this bug, and previous versions have stopped working.

**Comment by datakurre** [comment]
@LiamBindle Thank you for your environment file. I can confirm that I got the same issues with it.

But I was wondering, if you still had a working build from May-July, maybe you could find more comprehensive version list there? For example, file listing from `./xeus/kernel_packages` or contents of ` ./xeus/kernels/xpython/empack_env_meta.json` from the voici output directory?

I do share the gratitude and excitement for this project!

**Comment by LiamBindle** [comment]
@datakurre Ah good idea! Here is [empack_env_meta.json](https://github.com/user-attachments/files/17137584/empack_env_meta.json) from a deployment that looks like it's still working (note that alcesflow is our package).

Let me know if you have success pinning old versions.

**Comment by datakurre** [comment]
@LiamBindle Thank you. I did try those versions in `environment.yml`, but unfortunately the results were the same. Both `xeus-python = 0.16.0` and `0.17.1` failed similarly. Possibly I should try those versions with older Voici-releases later :thinking:

**Comment by LiamBindle** [comment]
@datakurre Thanks for trying. Your findings are consistent with mine. It seems like the existing binaries changed somehow, or there some browser thing broke the old packages. I did try an older version of voici but got the same result. I'm confused too

**Comment by datakurre** [comment]
Is this related to the issue fixed for pyodide-kernel in https://github.com/jupyterlite/pyodide-kernel/pull/148/files ?

I tried to build a project with pyodide-kernel instead, but failed with a completely unrelated build issue

```
error    libmamba Could not solve for environment specs
    The following package could not be installed
    └─ jupyterlite-pyodide-kernel 0.4.7**  is not installable because it requires
       └─ jupyterlite-core >=0.4.5,<0.5.0 , which requires
          └─ jupyter_core >=4.7  but there are no viable options
             ├─ jupyter_core 5.7.2 would require
             │  └─ __unix, which is missing on the system;
             └─ jupyter_core 5.7.2 would require
                └─ __win, which is missing on the system.
 ```

**Comment by jtpio** [comment]
@datakurre yeah it looks like a different issue when resolving the build environment. Feel free to open a different issue or discussion if needed, thanks!

**Comment by datakurre** [comment]
> Is this related to the issue fixed for pyodide-kernel in https://github.com/jupyterlite/pyodide-kernel/pull/148/files ?

I've been slowly learning to hack jupyterlite-xeus, but would anyone have ideas, what should I look into or try with it?

If I understand correctly, in pyodide-kernel a similar issue was fixed, but I still don't understand if that fix could be applied to jupyterlite-xeus (due to different architecture) or if this is even the same issue.

I assume that `jupyter.widget.control` related warnings are just symptoms of the original issue.

It is also still a mystery, why we are not able to find old set working versions :thinking: 

(My libmamba issue was just me.)

**Comment by datakurre** [comment]
> I was thinking that https://github.com/deathbeds/jupyterlite-pyodide-lock/ might help to load ipywidgets (or other packages) in the pyodide kernel without using await/micropip, would work like the xeus kernel where the packages are included.

That sounds interesting. Thanks. I assume that it would require some customization into voici build, it could still be easier to resolve than this mystery issue :thinking:

**Comment by jtpio** [comment]
> Until we are able to get Voici working with ipywidgets again, is there ways to may JupyterLite look more alike it: only show the notebook, hide code cells, auto run on load?

It should technically be possible (maybe even via a JupyterLab extension), but something to track in a separate issue?

**Comment by datakurre** [comment]
> > Until we are able to get Voici working with ipywidgets again, is there ways to may JupyterLite look more alike it: only show the notebook, hide code cells, auto run on load?
> 
> It should technically be possible (maybe even via a JupyterLab extension), but something to track in a separate issue?

Yes! I was expecting that such extensions would already exist, but if not, let's discuss it at https://github.com/jupyterlite/jupyterlite/issues/1556

**Comment by datakurre** [comment]
Ok. I learned my lesson to not try reinvent the wheel, but stay with voici.

I got voici working again with pyodide-kernel, thanks to Nick Bollweg, with:

* voici with fixed pyodide-kernel support https://github.com/voila-dashboards/voici/pull/138
* https://github.com/deathbeds/jupyterlite-pyodide-lock to lock and prefetch ipywidgets and other dependencies

This was still with patched voici 0.7.1 and jupyterlite 0.4.5. I have not tried the latest versions yet. I first created a jupyterlite project with pyodide-lock and once my notebooks were working, I added patched voici and built with `voici build`.
